### PR TITLE
optional download/previewer inputs - reporter

### LINKED ISSUE #26: custom yaml header fields in the Download Modules
The developer should have an option to specify which yaml header fields should be ask to be filled by the end user.
We are talking about fields like author, output, etc.

shiny inputs connected with yaml header fields should be named the same, input$author for author field ... . 
Should be possible after https://github.com/insightsengineering/teal.reporter/issues/11 as there will be a function to process the flat list of yaml header fields. 

We will add an argument to Download module with a vector of yaml files which should be taken into account.

UI for Download Button and Previewer have to be updated.

### PR DESCRIPTION
**Pull request by Polkas**
closes #26 

- support custom subset of inputs for download and previewer modules
- add additional checkmate validation
- additional tests

Example app with different subset of inputs (`rmd_yaml_args`) for download and previewer modules:

```
ui <- fluidPage(
  # please, specify specific bootstrap version and theme
  theme = bslib::bs_theme(version = "4"),
  titlePanel(""),
  tabsetPanel(
    tabPanel(
      "main App",
      tags$br(),
      sidebarLayout(
        sidebarPanel(
          uiOutput("encoding")
        ),
        mainPanel(
          tabsetPanel(
            id = "tabs",
            tabPanel("Plot", plotOutput("dist_plot")),
            tabPanel("Table", verbatimTextOutput("table")),
            tabPanel("Table DataFrame", verbatimTextOutput("table2")),
            tabPanel("Table DataTable", DT::dataTableOutput("table3"))
          )
        )
      )
    ),
    ### REPORTER
    tabPanel(
      "Previewer",
      reporter_previewer_ui("prev")
    )
    ###
  )
)
server <- function(input, output, session) {
  output$encoding <- renderUI({
    tagList(
      ### REPORTER
      teal.reporter::simple_reporter_ui("simple_reporter"),
      ###
      if (input$tabs == "Plot") {
        sliderInput(
          "binwidth",
          "binwidth",
          min = 2,
          max = 10,
          value = 8
        )
      } else if (input$tabs %in% c("Table", "Table DataFrame", "Table DataTable")) {
        selectInput(
          "stat",
          label = "Statistic",
          choices = c("mean", "median", "sd"),
          "mean"
        )
      } else {
        NULL
      }
    )
  })
  plot <- reactive({
    req(input$binwidth)
    x <- mtcars$mpg
    ggplot2::ggplot(data = mtcars, ggplot2::aes(x = mpg)) +
      ggplot2::geom_histogram(binwidth = input$binwidth)
  })
  output$dist_plot <- renderPlot(plot())

  table <- reactive({
    req(input$stat)
    lyt <- basic_table() %>%
      split_rows_by("Month", label_pos = "visible") %>%
      analyze("Ozone", afun = eval(str2expression(input$stat)))
    build_table(lyt, airquality)
  })
  output$table <- renderPrint(table())

  table2 <- reactive({
    req(input$stat)
    data <- aggregate(
      airquality[, c("Ozone"), drop = FALSE], list(Month = airquality$Month), get(input$stat),
      na.rm = TRUE
    )
    colnames(data) <- c("Month", input$stat)
    data
  })
  output$table2 <- renderPrint(print.data.frame(table2()))
  output$table3 <- DT::renderDataTable(table2())

  ### REPORTER
  reporter <- teal.reporter::Reporter$new()
  card_fun <- function(card = ReportCard$new(), comment) {
    if (input$tabs == "Plot") {
      card$set_name("Plot Module")
      card$append_text("My plot", "header2")
      card$append_plot(plot())
      card$append_rcode(
        paste(
          c(
            "x <- mtcars$mpg",
            "ggplot2::ggplot(data = mtcars, ggplot2::aes(x = mpg)) +",
            paste0("ggplot2::geom_histogram(binwidth = ", input$binwidth, ")")
          ),
          collapse = "\n"
        ),
        echo = TRUE,
        eval = FALSE
      )
    } else if (input$tabs == "Table") {
      card$set_name("Table Module rtables")
      card$append_text("My rtables", "header2")
      card$append_table(table())
      card$append_rcode(
        paste(
          c(
            "lyt <- rtables::basic_table() %>%",
            'rtables::split_rows_by("Month", label_pos = "visible") %>%',
            paste0('rtables::analyze("Ozone", afun = ', input$stat, ")"),
            "rtables::build_table(lyt, airquality)"
          ),
          collapse = "\n"
        ),
        echo = TRUE,
        eval = FALSE
      )
    } else if (input$tabs %in% c("Table DataFrame", "Table DataTable")) {
      card$set_name("Table Module DF")
      card$append_text("My Table DF", "header2")
      card$append_table(table2())
      # Here r code added as a regular verbatim text
      card$append_text(
        paste0(
          c(
            'data <- aggregate(airquality[, c("Ozone"), drop = FALSE], list(Month = airquality$Month), ',
            input$stat,
            ", na.rm = TRUE)\n",
            'colnames(data) <- c("Month", ', paste0('"', input$stat, '"'), ")\n",
            "data"
          ),
          collapse = ""
        ), "verbatim"
      )
    }
    if (!comment == "") {
      card$append_text("Comment", "header3")
      card$append_text(comment)
    }
    card
  }
  teal.reporter::simple_reporter_srv("simple_reporter",
    reporter = reporter, card_fun = card_fun,
    rmd_yaml_args = list(
      output = "html_document"
    )
  )
  teal.reporter::reporter_previewer_srv("prev", reporter,
    rmd_yaml_args = list(
      output = "html_document",
      toc = FALSE
    )
  )
  ###
}

if (interactive()) shinyApp(ui = ui, server = server)
```

### CODE REVIEW

File: R/DownloadModule.R [hunk 1]
```diff
@@ -39,7 +39,8 @@ download_report_button_ui <- function(id) {
 #' will appear in the `UI`.
 #' @param rmd_yaml_args `named list` vector with `Rmd` `yaml` header fields and their default values.
 #' Default `list(author = "NEST", title = "Report", date = Sys.Date(), output = "html_document", toc = FALSE)`.
-#' Please update only values at this moment.
+#' The `list` must include at least "output" field.
+#' The default value for `"output"` has to be in the `rmd_output` vector.
```
Comment (gogonzo) [inline R/DownloadModule.R:43]:
```suggestion
#' The `list` must include at least "output" field.
#' The default value for `"output"` needs to be an element of `rmd_output` vector.
```

File: R/SimpleReporter.R [hunk 2]
```diff
@@ -46,14 +46,36 @@ simple_reporter_ui <- function(id) {
...
 #' @param card_fun `function` which returns a [`ReportCard`] instance,
 #' the function has a `card` argument and an optional `comment` argument.
+#' @param rmd_output `character` vector with `rmarkdown` output types,
+#' by default all possible `c("pdf_document", "html_document", "powerpoint_presentation", "word_document")`.
+#' `pdflatex` has to be installed to use the `"pdf_document"` output. If vector is named then the names
+#' will appear in the `UI`.
+#' @param rmd_yaml_args `named list` vector with `Rmd` `yaml` header fields and their default values.
+#' Default `list(author = "NEST", title = "Report", date = Sys.Date(), output = "html_document", toc = FALSE)`.
+#' The `list` must include at least "output" field.
+#' The default value for `"output"` has to be in the `rmd_output` vector.
```
Comment (gogonzo) [inline R/SimpleReporter.R:56]: This documentation element is duplicated. Also `rmd_yaml_args` doesn't say anything about creation of the corresponding ui element via `reporter_download_inputs`. I think we should have one documentation entry somewhere (reporter_download_inputs or dummy function documentation) and inherit from there.

**Review by gogonzo — COMMENTED** [review]
Changes correct. Everything clear and working. I have a comment regarding documentation of `rmd_yaml_args` and `rmd_output`. It's not obvious that they hold instructions to create a corresponding UI elements. I'd highlight this in the docs. Also documentation of these arguments are duplicated in two places as they are passed to the same function `reporter_download_inputs` - I think this function should be documented and parameters should be inherited

File: R/SimpleReporter.R (hunk 2 above)
Comment (Polkas) [reply R/SimpleReporter.R:56]: Updated

[the remark this answers is not in the document]
Comment (Polkas) [reply R/SimpleReporter.R]: good catch:)

File: R/SimpleReporter.R [hunk 3]
```diff
@@ -46,14 +46,29 @@ simple_reporter_ui <- function(id) {
...
+                                )) {
   shiny::moduleServer(
     id,
     function(input, output, session) {
       add_card_button_srv("add_report_card_simple", reporter = reporter, card_fun = card_fun)
-      download_report_button_srv("download_button_simple", reporter = reporter)
+      download_report_button_srv(
+        "download_button_simple",
+        reporter = reporter, rmd_output = rmd_output, rmd_yaml_args = rmd_yaml_args
+      )
```
Comment (Polkas) [inline R/SimpleReporter.R:71]:
```suggestion
        reporter = reporter, 
        rmd_output = rmd_output, 
        rmd_yaml_args = rmd_yaml_args
      )
```

File: tests/testthat/test-DownloadReportModule.R [hunk 4]
```diff
@@ -35,6 +35,50 @@ testthat::test_that("download_report_button_srv - render and downlaod a document
   )
 })
 
+testthat::test_that("download_report_button_srv - subset of rmd_yaml_args", {
+  rmd_yaml_args_correct <- list(
```
Comment (Polkas) [inline tests/testthat/test-DownloadReportModule.R:39]:
```suggestion

  rmd_yaml_args_correct <- list(
```

**Review by gogonzo — APPROVED** [review]
👍

**Pull request merged**